lib/Models/ExtendModel.ts

```diff
@@ -0,0 +1,179 @@
+import { action, computed, observable, runInAction } from "mobx";
+import { getObjectId } from "../Traits/ArrayNestedStrataMap";
+import { ModelId } from "../Traits/ModelReference";
+import ModelTraits from "../Traits/ModelTraits";
+import { ObjectArrayTrait } from "../Traits/objectArrayTrait";
+import TraitsConstructor from "../Traits/TraitsConstructor";
+import addModelStrataView from "./addModelStrataView";
+import createStratumInstance from "./createStratumInstance";
+import { isLoadableStratum } from "./LoadableStratum";
+import ModelType, {
+  ArrayElementTypes,
+  BaseModel,
+  ModelConstructor,
+  ModelInterface
+} from "./Model";
+import StratumFromTraits from "./StratumFromTraits";
+import StratumOrder from "./StratumOrder";
+import Terria from "./Terria";
+
+/**
+ * A mixin to extend a model with new traits.
+ *
+ * @param Base The model to extend
+ * @param Traits The traits class for the extended model.
+ *        This should be a mix of the new traits and the  base class traits (and not just
+ *        the new traits).
+ * @return The extended model with all the `Traits` and the methods from base class.
+ */
+export default function ExtendModel<
+  T extends TraitsConstructor<ModelTraits>,
+  BaseClass extends ModelConstructor<BaseModel>
+>(
+  Base: BaseClass,
+  Traits: T
+): ModelConstructor<InstanceType<BaseClass> & ModelType<InstanceType<T>>> {
+  type Traits = InstanceType<T>;
+  type StratumTraits = StratumFromTraits<Traits>;
+
+  const MixinBase = Base as new (...a: any[]) => BaseModel;
+
```

Comment:
The rest of the code is mostly the same as the previous `CreateModel` impelmentation.